first gambling commit

ludvigsandberg opened the pull request.

ludvigsandberg committed:
first gambling commit

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+            await ctx.send('You must enter a positive number greater than 0.', ephemeral=True)
+            return
+        if amount > user_database_get('money', ctx.author.id):
+            await ctx.send('You don\'t have enough money.', ephemeral=True)
```
Rather than escaping you can use `"You don't have enough money."`. We've done this pretty arbitrarily so far but ig for longer strings double quotes should be preferred

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+        user_database_check_if_user_exists_otherwise_add(user.id)
+
+        embed = discord.Embed(title=f'{user.display_name}\'s wallet', color=discord.Color.blurple())
+        embed.set_thumbnail(url='https://cdn.iconscout.com/icon/free/png-512/free-wallet-588-456730.png')
```
What's the license for that image? If you want we can add it as emoji or sticker so we don't have our bot link to something off-platform that's also entirely out of our control

Tsche commented on requirements.txt:
```diff
@@ -2,3 +2,4 @@ fuzzywuzzy==0.18.0
 discord.py==2.1.0
 aiohttp==3.8.5
 python-Levenshtein==0.12.2
+datetime==5.4
```
Why do we need a package for that? Isn't python standard library datetime utils sufficient?

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+from src.util.user_database import *
+
+
+daily_amount = 10
```
please put configuration options in config.py

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+        """
+        Place a bet (50/50). Enter the amount to bet.
+        """
+        user_database_check_if_user_exists_otherwise_add(ctx.author.id)
```
that's one long af identifier

YuhanunCitgez replied:
user_database_upsert is better

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+from discord import Member
+from datetime import datetime
+
+from src.util.user_database import *
```
don't use wildcard imports

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+            await ctx.send('You must enter a positive number greater than 0.', ephemeral=True)
+            return
+        if amount_to_bet > amount_in_wallet:
+            await ctx.send('You don\'t have enough money.', ephemeral=True)
```
escape can be gotten rid of by using double quotes

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+            await ctx.send('You don\'t have enough money.', ephemeral=True)
+            return
+
+        won = random.randint(1, 2) == 1
```
`random.choice([True, False])` or `bool(random.getrandbits(1))` should be preferred

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+        won = random.randint(1, 2) == 1
+
+        # Update wallet.
+        amount_result = amount_in_wallet + (amount_to_bet if won else -amount_to_bet)
```
I really dislike `(amount_to_bet if won else -amount_to_bet)`. I think it's better to be explicit here, even though it's a little more verbose

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+
+        last_daily_timestamp = int(user_database_get('last_daily_timestamp', ctx.author.id))
+        current_timestamp = int(time.time())
+        seconds_in_a_day = 86400
```
true constants should be at module level and ALL_CAPS

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+    @commands.hybrid_command(with_app_command=True)
+    async def give(self, ctx: Context, user: Member, amount: int):
+        """
+        Transfer money to another member. First argument is user ID. Ssecond argument is amount.
```
`Ssecond` typo

Tsche commented on src/cogs/gamble.py:
```diff
@@ -0,0 +1,152 @@
+        # Poor man's clamp.
+        entries = max(1, min(entries, 50))
+
+        query = user_database.execute(f'SELECT id, money FROM users ORDER BY money DESC LIMIT ?', [entries])
```
you have extracted all database queries except for this one, why?

Tsche commented on src/util/user_database.py:
```diff
@@ -0,0 +1,31 @@
+import sqlite3
+
+user_database_connection = sqlite3.connect('users.db')
```
database path should be configurable (in config.py)

YuhanunCitgez replied:
This file is non-persistent

Tsche commented on src/util/user_database.py:
namespacing by prefixing everything with `user_database` is yucky. Please wrap it in a class instead (or drop the prefix entirely and import the module instead of its contents to force usage of its name).

I don't like how it does a lot of stuff when importing the module (in Python modules are essentially singletons). That should not be required and will make testing a lot harder if we ever decide to write tests for any of this. If you want to ensure there's exactly one instance of this, you should probably open up the database connection during bot initialization instead.

Tsche commented on src/util/user_database.py:
```diff
@@ -0,0 +1,31 @@
+    count = query.fetchone()[0]
+
+    # If no user ID was found in database, add user.
+    if count == 0:
```
this should be `if not count`

ludvigsandberg committed:
gambling related improvements

Still need to fix /daily and turn the wallet icon into a server sticker.

ludvigsandberg committed:
Fixed /daily and sticker in embed

Still need to update the sticker id in config.py after Che uploads the sticker.

ludvigsandberg committed:
Fixed all of Che's requests.

YuhanunCitgez commented:
@Tsche we still wanna move ahead?

Tsche commented:
> @Tsche we still wanna move ahead?

I'm currently super busy but in general yeah sure why not. Not sure what the current state of this PR is, I'll try to look into it some time this week.

makian123 commented:
Is the input properly sanitized, since its interacting with database (amount to gamble cant be a string and where it accepts string it cant be like `;drop table users`)

bu420 commented:
@makian123 SQL injections won't be a problem as long as we use parameterized queries with question marks instead of raw string operations

makian123 commented:
Alright, that is pretty cool

makian123 commented:
you need to modify the docker run args so it takes in a volume, that way sqlite database will be persistent

ludvigsandberg committed:
Moved database.

Commit:
persistent volume

Commit:
persistent volume

ludvigsandberg committed:
Merge pull request #2 from bizxeon/master

Pesistent Docker Volume

makian123 commented:
I have one big issue with this, your database is non scalable, the class you've written in utils will only work for the gambling stuff, can you generalize the DB class so it uses tables and you could set field in a custom way? That way you can just write a wrapper for your gambling.
Requesting this because i have some moderation feature plans in future and it'd make it easier to have one sqlite database with different tables

ludvigsandberg committed:
Moved gamble specific functions from db.py

bu420 commented:
@makian123 moved gambling specific functions from db.py into gamble.py.
